### PR TITLE
feat: create DefaultChargeLinks

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description

This PR will add a domain model for `DefaultChargeLink`, and the ability to read it from the database.

Note that I have updated where we map `DateTimes` we now recast the to a new `DateTime` with `DateTime.Kind` being `UTC`

## References

* #476

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        /// <summary>
+        /// The date the charge is applicable from.
+        /// </summary>
+        public Instant ApplicableDate { get; set; }
+
+        /// <summary>
+        /// The default charge.
+        /// </summary>
+        public Charge DefaultCharge { get; set; }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs:31]: I'm afraid that I concurred with this earlier. I'm, however, concerned about the consequences. For example what should you instantiate in order to create and persist a new default charge link? Can/may/must the `DefaultCharge` have value or be null?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs [hunk 2]
```diff
@@ -40,6 +40,8 @@ public ChargesDatabaseContext(DbContextOptions<ChargesDatabaseContext> options)
 
         public DbSet<MeteringPoint> MeteringPoints { get; set; }
 
+        public DbSet<DefaultChargeLinkSetting> DefaultChargeLinkSetting { get; set; }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs:43]: Please use plural form.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs (hunk 1 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs:31]: Let's have a call about this when time allows.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using NodaTime;
+
+#pragma warning disable 8618
+
+namespace GreenEnergyHub.Charges.Domain.Charges
+{
+    public class DefaultChargeLinkSetting
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs:21]: Does "setting" everywhere actually add anything but longer names to "default charge link"?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public async Task<IEnumerable<DefaultChargeLinkSetting>> GetDefaultChargeLinkSettingAsync(MeteringPointType meteringPointType)
+        {
+            var defaultChargeLinkSettings = await _chargesDatabaseContext.DefaultChargeLinkSetting
+                .Include(x => x.Charge)
+                .Include(x => x.Charge.ChargeOperation)
+                .Include(x => x.Charge.ChargePeriodDetails)
+                .Include(x => x.Charge.MarketParticipant)
+                .Where(x => x.MeteringPointType == (int)meteringPointType).ToListAsync().ConfigureAwait(false);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs:46]: Is `foo.Charge.ChargeOperation.ChargePeriodDetails == foo.Charge.ChargePeriodDetails`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            _chargesDatabaseContext = chargesDatabaseContext;
+        }
+
+        public async Task<IEnumerable<DefaultChargeLinkSetting>> GetDefaultChargeLinkSettingAsync(MeteringPointType meteringPointType)
+        {
+            var defaultChargeLinkSettings = await _chargesDatabaseContext.DefaultChargeLinkSetting
+                .Include(x => x.Charge)
+                .Include(x => x.Charge.ChargeOperation)
+                .Include(x => x.Charge.ChargePeriodDetails)
+                .Include(x => x.Charge.MarketParticipant)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs:45]: This adds a high risk that extensions to the Charge model is not reflected here and the user of the repository will end up with an incomplete charge.

Comment (HenrikSommer) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs:45]: Using `.Include` will make Entity Framework load all properties of all included types. Do we really need all properties of all types?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            var actual = await
+                sut.GetDefaultChargeLinkSettingAsync(MeteringPointType.Consumption).ConfigureAwait(false);
+            var actualDefaultChargeLinkSettings = actual as DefaultChargeLinkSetting[] ?? actual.ToArray();
+
+            // Assert
+            actualDefaultChargeLinkSettings.Should().NotBeNullOrEmpty();
+            actualDefaultChargeLinkSettings.First().DefaultCharge.Id.Should().BeEquivalentTo(ExpectedChargeId);
+        }
+
+        private void SeedDatabase()
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs:60]: How about `CreateAndSeedDatabase`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            // Arrange
+            SeedDatabase();
+            await using var chargesDatabaseContext = new ChargesDatabaseContext(_dbContextOptions);
+            var sut = new DefaultChargeLinkSettingRepository(chargesDatabaseContext);
+
+            // Act
+            var actual = await
+                sut.GetDefaultChargeLinkSettingAsync(MeteringPointType.Consumption).ConfigureAwait(false);
+            var actualDefaultChargeLinkSettings = actual as DefaultChargeLinkSetting[] ?? actual.ToArray();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs:53]: I think this ought to be moved to the assert section.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs (hunk 5 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs:45]: Yes they are used for mapping the Charge

Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs:45]: I think you are on to something here @HenrikSommer 😄 . Perhaps we don't even need to map to domain charge.

For creating a `ChargeLinkCommand` we need `ChargeId`, `ChargeOwner`, `ChargeType`. Those can be retrieved from tables `Charge` and MarketParticipant`

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs (hunk 3 above)
Comment (prtandrup) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs:21]: Although I just said something else to Lasse on the phone, I think we should aim for DefaultChargeLink.cs and refactor the props to no longer contain a (domain) charge, but instead the required data to create a ChargeLinkCommand.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs (hunk 5 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs]: We have discussed it. And the PR should now represent the idea on how to reference a Charge.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs (hunk 3 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLinkSetting.cs]: Think we went over this, ill mark it as resolved.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs (hunk 5 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkSettingRepository.cs]: This has been changed, we no longer include anything from the Charge in this repo.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        public Instant ApplicableDate { get; set; }
+
+        /// <summary>
+        /// The date the charge is no longer applicable for linking.
+        /// </summary>
+        public Instant EndDate { get; set; }
+
+        /// <summary>
+        /// Reference to the Charge the DefaultChargeLink.
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:35]: Change to something like this?
```suggestion
        /// A reference to the charge in the Charge table
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+#pragma warning disable 8618
+
+namespace GreenEnergyHub.Charges.Domain.Charges
+{
+    // Logically there is a MeteringPointType attached to the DefaultChargeLink, but atm. It is not used.
+    public class DefaultChargeLink
+    {
+        /// <summary>
+        /// The date the charge is applicable for linking.
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:25]: "Must be greater than or equal to..."

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        public Instant ApplicableDate { get; set; }
+
+        /// <summary>
+        /// The date the charge is no longer applicable for linking.
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:30]: "Must be less than or equal to..."

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/Model/DefaultChargeLinkSetting.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        [ForeignKey("Charge")]
+        public int ChargeRowId { get; set; }
+
+        public int MeteringPointType { get; set; }
+
+        public DateTime StartDateTime { get; set; }
+
+        public DateTime? EndDateTime { get; set; }
+
+        public virtual Charge Charge { get; set; }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/Model/DefaultChargeLinkSetting.cs]: Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    public class DefaultChargeLinkRepository : IDefaultChargeLinkRepository
+    {
+        private readonly IChargesDatabaseContext _chargesDatabaseContext;
+
+        public DefaultChargeLinkRepository(IChargesDatabaseContext chargesDatabaseContext)
+        {
+            _chargesDatabaseContext = chargesDatabaseContext;
+        }
+
+        public async Task<IEnumerable<DefaultChargeLink>> GetDefaultChargeLinksAsync(MeteringPointType meteringPointType)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs:38]: How about `GetAsync`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            var defaultChargeLinkSettings = await _chargesDatabaseContext.DefaultChargeLinkSettings
+                .Where(x => x.MeteringPointType == (int)meteringPointType).ToListAsync().ConfigureAwait(false);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs:41]: I'm personally very much in favour of fluent expressions on either a single line or with each dot starting a new line.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            return new DefaultChargeLink
+            {
+                ApplicableDate = Instant.FromDateTimeUtc(defaultChargeLinkSettings.StartDateTime.ToUniversalTime()),
+                EndDate = defaultChargeLinkSettings.EndDateTime != null ?
+                    Instant.FromDateTimeUtc(defaultChargeLinkSettings.EndDateTime.Value.ToUniversalTime()) :
+                    Instant.MinValue,
+                ChargeRowId = defaultChargeLinkSettings.ChargeRowId,
+            };
+        }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs:56]: SOLID principles would suggest this being segregated to a (data) mapper.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+            using var context = new ChargesDatabaseContext(_dbContextOptions);
+            context.Database.EnsureDeleted();
+            context.Database.EnsureCreated();
+            var marketParticipant = context.MarketParticipants.Add(
+                new MarketParticipant { Name = "Name", Role = 1, MarketParticipantId = "MarketParticipantId" });
+            context.SaveChanges();
+            var charge = context.Charges.Add(new Charge
+            {
+                Currency = "DKK",
+                Resolution = 1,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/DefaultChargeLinkSettingRepositoryTests.cs:77]: I have no clue of any of all these integer values. How about replacering all like this:

```suggestion
                Resolution = (int)Resolution.PT15M,
```

**Review by BjarkeMeier — COMMENTED** [review]
Much better IMHO :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs (hunk 9 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:25]: @prtandrup advised it should only be greater than.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs (hunk 10 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:36]: We now expose a property to expose if the DefaultChargeLink is applicable for linking or not.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+        /// <summary>
+        /// If the charge setting has a EndDateTime, it is only applicable for link if it is lesser or equal too StartDateTime.
+        /// </summary>
+        public bool ApplicableForLinking => _endDateTime == null || _endDateTime.Value >= StartDateTime;
+
+        /// <summary>
+        /// A reference to the charge in the Charge table
+        /// </summary>
+        public int ChargeRowId { get; }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:67]: This has nothing to do with your PR, but I think we should reconsider (in another story) the naming regarding "RowId" as it is very persistence-oriented.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+namespace GreenEnergyHub.Charges.Domain.Charges
+{
+    // Logically there is a MeteringPointType attached to the DefaultChargeLink, but atm. It is not used.
+    public class DefaultChargeLink
+    {
+        /// <summary>
+        /// The date the MeteringPoint is created. And the date the charge link should be applied from
+        /// if it is greater than the default charge link StartDateTime.
+        /// </summary>
+        private readonly Instant _meteringPointCreatedDateTime;
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:28]: I love that you're starting to work with the domain models!
I, however, think that we need to tweek it a little as `_meteringPointCreatedDateTime` adds a hidden state that is not evident for users of the default charge link. It actually tailors this default link to a specific date, which is in contrast to the concept of what a default charge link is. So you should rather have this date as an to a `GetStartDateTime()` method (and removing the `StartDateTime` prop).

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/Model/DefaultChargeLinkSetting.cs [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+#pragma warning disable 8618
+
+namespace GreenEnergyHub.Charges.Infrastructure.Context.Model
+{
+    public class DefaultChargeLinkSetting
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/Model/DefaultChargeLinkSetting.cs:23]: I think it's a mistake to try and invent a distinction between default link setting and default setting. I think there should only be a default charge link (with a domain and a data model representation).

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+        public async Task<IEnumerable<DefaultChargeLink>> GetAsync(MeteringPointType meteringPointType, Instant meteringPointCreatedDateTime)
+        {
+            var defaultChargeLinkSettings = await _chargesDatabaseContext.DefaultChargeLinkSettings
+                .Where(x => x.MeteringPointType == (int)meteringPointType)
+                .ToListAsync()
+                .ConfigureAwait(false);
+
+            return defaultChargeLinkSettings.Select(x => DefaultChargeLinkMapper.Map(meteringPointCreatedDateTime, x)).ToList();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Repositories/DefaultChargeLinkRepository.cs:44]: As per other comment I think that we should not have a setting. But if you end up having it then this is probably rather a factory than a (data) mapper.

**Review by prtandrup — COMMENTED** [review]
A few comments for now. Still need to look at the last couple of test files.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs (hunk 16 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:67]: Good idea - how about we talk about it in the team dev session?

Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:67]: Haha I can see you already created the topic :)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        /// The starting date is determined by the latest date when comparing meteringPointCreatedDateTime and
+        /// SettingStartDateTime. It is used to dictate when the charge link should start from.
+        /// </summary>
+        public Instant GetStartDateTime(Instant meteringPointCreatedDateTime)
+        {
+            return _startDateTime > meteringPointCreatedDateTime ? _startDateTime : meteringPointCreatedDateTime;
+        }
+
+        /// <summary>
+        /// If the charge setting has a EndDateTime, it is only applicable for link if it is lesser or equal too StartDateTime.
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/DefaultChargeLink.cs:63]:
```suggestion
        /// If the DefaultChargeLink has an EndDateTime, it is only applicable for linking when EndDateTime is later than the StartDateTime
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs [hunk 21]
```diff
@@ -55,6 +57,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             modelBuilder.Entity<Charge>().ToTable("Charge");
             modelBuilder.Entity<MarketParticipant>().ToTable("MarketParticipant");
             modelBuilder.Entity<MeteringPoint>().ToTable("MeteringPoint");
+            modelBuilder.Entity<DefaultChargeLink>().ToTable("DefaultChargeLinkSetting");
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargesDatabaseContext.cs:60]: After this [PR](https://github.com/Energinet-DataHub/geh-charges/pull/548/files) was merged, please change to this:
```suggestion
            modelBuilder.Entity<DefaultChargeLink>().ToTable("DefaultChargeLink");
```

**Review by HenrikSommer — APPROVED** [review]

**Review by prtandrup — APPROVED** [review]
Besides my two minor comments, LGTM 😄

**Pull request merged**